reshape and document the dht-cache

lu-zero opened the pull request:
- Reduce the public API
- Document it

lu-zero committed:
Document dht-cache

lu-zero committed:
Hide unneeded API surface

lu-zero committed:
Re-export to the root of dht-cache

lu-zero committed:
Initial error type

lu-zero committed:
Simplify the DomoEvent handling

lu-zero committed:
wip: Add get_peers_stats

lu-zero committed:
Hide even more of the DomoCache

lu-zero committed:
Move the data structures in a separate module

lu-zero committed:
Derive Default for DomoCacheElement

lu-zero committed:
Move the rsa key generator in domolibp2p

lu-zero committed:
Make the pg test use the DOMO_DHT_TEST_DB env var if available

lu-zero committed:
Split the behavior creation from the swarm setup

Should make easier testing the network component using
libp2p_swarm_test.

lu-zero committed:
wip: Abstract over the network code

lu-zero committed:
Fix documentation

lu-zero committed:
wip: Rework the cache layer

lu-zero committed:
wip: Update the broker to handle the NewPeers event

lu-zero committed:
fixup: use any() instead of find().is_some()

lu-zero committed:
fixup: Move the comments to the right place

lu-zero committed:
Wire in the persistent storage

lu-zero committed:
Allow to not use a local db at all

lu-zero committed:
fixup: Clippy lints

lu-zero committed:
Test everything including pg integration

lu-zero committed:
wip: Add Builder

lu-zero committed:
Refactor the psk logic

lu-zero committed:
fixme: move to dev-dependencies

lu-zero committed:
Use pnet in the ephemeral swarms to ensure tests do not interact due mdns

lu-zero committed:
Speed up the test execution

lu-zero committed:
Speedup the test execution

The resend task would not interfere with the rest this way.

lu-zero committed:
fixup: Spurious comment

lu-zero committed:
fixup: is_syncronised is not async

lu-zero committed:
Do not keep the write lock longer than needed

lu-zero committed:
Explicitly yield in the swarm task

lu-zero committed:
Compact the CI

lu-zero committed:
Add a mean to access the list of peers

lu-zero committed:
Use more workers in the tests

lu-zero committed:
fixme: Check for a subset of peers

lu-zero committed:
Add Cache::get_hash

dodomorandi commented on dht-cache/src/cache/local.rs:
```diff
@@ -0,0 +1,320 @@
+    /// Try to insert the element in the cache
+    ///
+    /// Return Err(()) if the element to insert is older than the one in the cache
+    pub async fn try_put(&self, elem: &DomoCacheElement) -> Result<(), ()> {
```
I don't like this approach, IMHO it is better to return a `bool` at this point and mark the function as `#[must_use]`. It should also make the implementation a bit more clear.

lu-zero replied:
I was on the fence between the two, given the usage the boolean should even simplify the code further.

dodomorandi commented on dht-cache/src/cache/local.rs:
```diff
@@ -0,0 +1,320 @@
+
+impl Query {
+    /// Create a new query over a local cache
+    pub fn new(topic: &str, cache: LocalCache) -> Self {
+        Self {
+            topic: topic.to_owned(),
```
```suggestion
    pub fn new(topic: impl Into<String>, cache: LocalCache) -> Self {
        Self {
            topic: topic.into(),
```

dodomorandi commented on dht-cache/src/cache/local.rs:
```diff
@@ -0,0 +1,320 @@
+        }
+    }
+    /// Look up for a specific uuid
+    pub fn with_uuid(mut self, uuid: &str) -> Self {
+        self.uuid = Some(uuid.to_owned());
```
```suggestion
    pub fn with_uuid(mut self, uuid: impl Into<String>) -> Self {
        self.uuid = Some(uuid.into());
```

dodomorandi commented on dht-cache/src/cache.rs:
```diff
@@ -0,0 +1,462 @@
+                    std::fs::write(pk_path, pem)?;
+                    der
+                }
+                Err(e) => Err(e)?,
```
```suggestion
                Err(e) => return Err(e.into()),
```

dodomorandi commented on dht-cache/src/cache.rs:
```diff
@@ -0,0 +1,462 @@
+    /// Send a volatile message
+    ///
+    /// Volatile messages are unstructured and do not persist in the DHT.
+    pub fn send(&self, value: &Value) -> Result<(), Error> {
+        self.cmd
+            .send(Command::Broadcast(value.to_owned()))
```
```suggestion
    pub fn send(&self, value: Value) -> Result<(), Error> {
        self.cmd
            .send(Command::Broadcast(value))
```

dodomorandi commented on dht-cache/src/cache.rs:
```diff
@@ -0,0 +1,462 @@
+    /// Persist a value within the DHT
+    ///
+    /// It is identified by the topic and uuid value
+    pub async fn put(&self, topic: &str, uuid: &str, value: &Value) -> Result<(), Error> {
+        let elem = DomoCacheElement {
+            topic_name: topic.to_string(),
+            topic_uuid: uuid.to_string(),
+            value: value.to_owned(),
```
```suggestion
    pub async fn put(&self, topic: impl Into<String>, uuid: impl Into<String>, value: Value) -> Result<(), Error> {
        let elem = DomoCacheElement {
            topic_name: topic.into(),
            topic_uuid: uuid.into(),
            value,
```

dodomorandi commented on dht-cache/src/cache.rs:
```diff
@@ -0,0 +1,462 @@
+    ///
+    /// It inserts the deletion entry and the entry value will be marked as deleted and removed
+    /// from the stored cache.
+    pub async fn del(&self, topic: &str, uuid: &str) -> Result<(), Error> {
+        let elem = DomoCacheElement {
+            topic_name: topic.to_string(),
+            topic_uuid: uuid.to_string(),
```
```suggestion
    pub async fn del(&self, topic: impl Into<String>, uuid: impl Into<String>) -> Result<(), Error> {
        let elem = DomoCacheElement {
            topic_name: topic.into(),
            topic_uuid: uuid.into(),
```

dodomorandi commented on dht-cache/src/cache.rs:
```diff
@@ -0,0 +1,462 @@
+    let cmd_update = cmd.clone();
+    let peer_id = local_peer_id.clone();
+
+    tokio::task::spawn(async move {
```
I think that this could be transformed into a `spawn_local` without any issue, leading to the possibility of removing the `unsafe impl std::marker::Sync for InnerCache {}`

lu-zero replied:
We can try that way, I have a pending patch that moves the writes in a separate task using a channel, but if we can get away with a LocaSet it sounds overall cleaner.

dodomorandi commented on dht-cache/src/cache.rs:
```diff
@@ -0,0 +1,462 @@
+                                let mut elem = elem.to_owned();
+                                log::debug!("resending {}", elem.topic_uuid);
+                                elem.republication_timestamp = utils::get_epoch_ms();
+                                cmd.send(Command::Publish(serde_json::to_value(&elem).unwrap()))
+                                    .unwrap();
```
```suggestion
                               
                                // This cannot fail because `cmd` is the sender part of the
                                // `stream` we are currently reading. In practice, we are
                                // queueing the commands in order to read them later.
                                cmd.send(Command::Publish(serde_json::to_value(&elem).unwrap()))
                                    .unwrap();
```

dodomorandi commented on dht-cache/src/cache.rs:
```diff
@@ -0,0 +1,462 @@
+        expected_peers.insert(b_c.peer_id.clone());
+        expected_peers.insert(c_c.peer_id.clone());
+
+        tokio::task::spawn(async move {
```
I think we should avoid spawning tasks in tests, and "just" joining them. Spawning tasks in the runtime could make edge cases work by chance that otherwise, if we are lucky enough, could more easily deadlock in single thread and be detected by tests.

lu-zero replied:
You are absolutely right.

dodomorandi commented on dht-cache/src/cache.rs:
```diff
@@ -0,0 +1,462 @@
+                    .await;
+            }
+
+            loop {
+                let (node, ev) = tokio::select! {
+                    v = a_ev.next() => ("a", v.unwrap()),
+                    v = b_ev.next() => ("b", v.unwrap()),
+                    v = c_ev.next() => ("c", v.unwrap()),
+                };
+
+                match ev {
+                    DomoEvent::PersistentData(data) => {
+                        log::debug!("{node}: Got data {data:?}");
+                    }
+                    _ => {
+                        log::debug!("{node}: Other {ev:?}");
+                    }
+                }
+            }
+        });
```
This is not cancel safe, and this is the reason we should avoid using `select!` when possible. This is a possible approach using `futures-concurrency`:

```rust
            (
                a_ev.map(|v| ("a", v)),
                b_ev.map(|v| ("b", v)),
                c_ev.map(|v| ("c", v)),
            )
                .merge()
                .for_each(|(node, ev)| async move {
                    match ev {
                        DomoEvent::PersistentData(data) => {
                            log::debug!("{node}: Got data {data:?}");
                        }
                        _ => {
                            log::debug!("{node}: Other {ev:?}");
                        }
                    }
                })
                .await;
```

lu-zero replied:
You are right, the stream accesses RwLocks, it should be also documented

dodomorandi reviewed (changes requested):
I still need to finish reviewing, I have two major concerns about use of `unsafe` and `select!` other things are minor.

lu-zero committed:
Apply the impl Into<String>/owned Value pattern

Co-authored-by: Edoardo Morandi <morandidodo@gmail.com>

lu-zero committed:
Use a worker task to persist the data

lu-zero committed:
Use futures-concurrency

lu-zero committed:
Signal when the nodes are available, not when they are discovered

dodomorandi commented on dht-cache/src/cache/local.rs:
```diff
@@ -0,0 +1,320 @@
+
+            for (topic_uuid, value) in map_topic_name.iter() {
+                topic_uuid.hash(state);
+                value.to_string().hash(state);
```
Is there a way to easily split these comments into a separated issue? Obviously, for me it is easy to _just check the diff_, but I agree that the aim here is a bit different. If we can easily split issues that should be fixed in different PRs, it would be great.

lu-zero replied:
Sadly there isn't, it would be a nice feature though.

I'd do a round of performance improvements after we are good with the API rework :)

dodomorandi reviewed (changes requested):
Ok, just a major concern, most of the things are minor or considerations.

lu-zero committed:
Do not panic on non-utf8 data

lu-zero commented on dht-cache/src/cache/local.rs:
```diff
@@ -0,0 +1,320 @@
+
+/// Local cache
+#[derive(Default, Clone)]
+pub struct LocalCache(Arc<RwLock<InnerCache>>);
```
I started with simply moving the db interaction to a separate task, then we can see if having a localset can integrate with the rest of the system.

lu-zero commented on dht-cache/src/cache/local.rs:
```diff
@@ -0,0 +1,320 @@
+unsafe impl std::marker::Sync for InnerCache {}
+
+impl InnerCache {
+    pub fn put(&mut self, elem: &DomoCacheElement) {
```
We need to make the whole api uniform regarding taking reference or taking ownership, ergonomy-wise taking the reference feels nicer.

dodomorandi replied:
I don't agree, but do what you think it is the best for the project :relaxed:.

lu-zero committed:
use into_iter

lu-zero committed:
Cleanup the peers check to be more readable

lu-zero commented on dht-cache/src/dht.rs:
```diff
@@ -0,0 +1,369 @@
+                }
+            }
+        }
+
+        drop(b_s);
+        drop(c_s);
```
It is now :)

lu-zero committed:
drop redundant drops

lu-zero committed:
Avoid redundant path

lu-zero committed:
fixup: multiple_peer tests should wait for all of them

Otherwise it would infiniloop it send the volatile message when only one
of the two peers is ready.

lu-zero committed:
Take ownership of the elem in the local cache API as well

dodomorandi commented on dht-cache/src/dht.rs:
```diff
@@ -0,0 +1,390 @@
+                // the mdns event is not enough to ensure we can send messages
+                _ = interval.tick() => {
+                    log::debug!("{} Checking for peers", swarm.local_peer_id());
+                    let peers: Vec<_> = check_peers(&mut swarm);
+                        if !peers.is_empty() &&
+                        ev_send.send(Event::Ready(peers)).is_err() {
+                            return swarm;
+                    }
```
Formatting:
```suggestion
                    let peers: Vec<_> = check_peers(&mut swarm);
                    if !peers.is_empty() && ev_send.send(Event::Ready(peers)).is_err() {
                        return swarm;
                    }
```

lu-zero committed:
Refine the Ready peers event

dodomorandi committed:
fix: export structs from cache::local

dodomorandi committed:
feat!: overhaul dht_cache::cache::Query API

lu-zero committed:
Mark Cache as Clone
